File: java/hypertable-common/src/main/java/org/hypertable/FsBroker/Lib/MetricsHandler.java
/*
 * Copyright (C) 2007-2014 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/** @file
 * Declarations for MetricsHandler.
 * This file contains declarations for MetricsHandler, a dispatch handler class
 * used to collect and publish FsBroker metrics.
 */

package org.hypertable.FsBroker.Lib;

import java.lang.Integer;
import java.util.Properties;
import java.util.logging.Logger;
import org.hypertable.AsyncComm.Comm;
import org.hypertable.AsyncComm.DispatchHandler;
import org.hypertable.AsyncComm.Event;
import org.hypertable.Common.MetricsCollectorGanglia;
import org.hypertable.Common.MetricsProcess;

/** Collects and publishes FSBroker metrics.
 * This class acts as the timer dispatch handler for periodic metrics
 * collection in an FSBroker.
 */
public class MetricsHandler implements DispatchHandler {

  static final Logger log = Logger.getLogger("org.hypertable.FsBroker.Lib");

  /** Constructor.
   * Initializes #mCollectionInterval to the property
   * <code>Hypertable.Monitoring.Interval</code> and allocates a Ganglia
   * collector object, initializing it with "fsbroker" and
   * <code>Hypertable.Metrics.Ganglia.Port</code>.  Lastly, calls
   * Comm.SetTimer() to register a timer for
   * #mCollectionInterval milliseconds in the future and passes
   * <code>this</code> as the timer handler.
   * @param comm Comm layer object
   * @param props %Properties object
   */
  public MetricsHandler(Comm comm, Properties props) {
    String str;

    str = props.getProperty("Hypertable.Metrics.Ganglia.Port", "15860");
    int port = Integer.parseInt(str);

    mMetricsCollectorGanglia = new MetricsCollectorGanglia("fsbroker", port);
    mMetricsCollectorGanglia.update("type", "hadoop");

    str = props.getProperty("Hypertable.Monitoring.Interval", "30000");
    mCollectionInterval = Integer.parseInt(str);

    mComm = comm;

    mLastTimestamp = System.currentTimeMillis();

    mComm.SetTimer(mCollectionInterval, this);

  }

  /** Collects and publishes metrics.
   * This method updates the <code>type</code>, <code>errors</code>,
   * <code>syncs</code>, <code>syncLatency</code>, <code>readThroughput</code>,
   * <code>writeThroughput</code> and general process metrics and publishes them
   * via mMetricsCollectorGanglia.  After metrics have been collected, the
   * timer is re-registered for mCollectionInterval milliseconds in the future
   * and the metrics collection member variables are reset.
   * @param event Comm layer TIMER event
   */
  public void handle(Event event) {

    if (event.type == Event.Type.TIMER) {
      long now = System.currentTimeMillis();
      long elapsed_millis = now - mLastTimestamp;
      long elapsed_seconds = elapsed_millis / 1000;
      mMetricsProcess.collect(now, mMetricsCollectorGanglia);
      mMetricsCollectorGanglia.update("type", "hadoop");
      synchronized (this) {
        mMetricsCollectorGanglia.update("errors", mErrors);
        double sps = (double)mSyncs / (double)elapsed_seconds;
        mMetricsCollectorGanglia.update("syncs", sps);
        int avgSyncLatency = (mSyncs > 0) ? mSyncLatency/mSyncs : 0;
        mMetricsCollectorGanglia.update("syncLatency", avgSyncLatency);
        if (elapsed_millis > 0) {
          long mbps = (mBytesRead / 1000000) / elapsed_seconds;
          mMetricsCollectorGanglia.update("readThroughput", (int)mbps);
          mbps = (mBytesWritten / 1000000) / elapsed_seconds;
          mMetricsCollectorGanglia.update("writeThroughput", (int)mbps);
        }
        mLastTimestamp = now;
        mErrors = 0;
        mSyncs = 0;
        mSyncLatency = 0;
        mBytesRead = 0;
        mBytesWritten = 0;
      }
      try {
        mMetricsCollectorGanglia.publish();
      }
      catch (Exception e) {
        log.info("Problem publishing Ganglia metrics - " +
                 e.getMessage());
      }
    }
    else {
      System.out.println("Unexpected event - " + event);
      System.exit(-1);
    }

    mComm.SetTimer(mCollectionInterval, this);    
  }

  /** Adds bytes read.
   * Adds <code>count</code> to mBytesRead.
   * @param count Count of bytes read
   */
  public synchronized void addBytesRead(long count) {
    mBytesRead += count;
  }

  /** Adds bytes written.
   * Adds <code>count</code> to mBytesWritten.
   * @param count Count of bytes written
   */
  public synchronized void addBytesWritten(long count) {
    mBytesWritten += count;
  }

  /** Adds sync information.
   * Adds <code>latency</code> to mSyncLatency and increments mSyncs.
   * @param latency Latency of sync
   */
  public synchronized void addSync(long latency) {
    mSyncs++;
    mSyncLatency += (int)latency;
  }

  /** Increments error count.
   * Increments mErrors.
   */
  public synchronized void incrementErrorCount() {
    mErrors++;
  }

  /** Metrics collection interval */
  private int mCollectionInterval = 0;

  /** Comm layer object */
  private Comm mComm;

  /** Process metrics tracker */
  private MetricsProcess mMetricsProcess = new MetricsProcess();

  /** Ganglia metrics collector */
  private MetricsCollectorGanglia mMetricsCollectorGanglia;

  /** Timestamp (ms) of last metrics collection */
  private long mLastTimestamp;

  /** Bytes written since last metrics collection */
  private long mBytesWritten = 0;

  /** Bytes read since last metrics collection */
  private long mBytesRead = 0;

  /** Cumulative sync latency since last metrics collection */
  private int mSyncLatency = 0;

  /** Syncs since last metrics collection */
  private int mSyncs = 0;

  /** Error count since last metrics collection */
  private long mErrors = 0;

}

File: src/cc/FsBroker/Lib/MetricsHandler.cc
/*
 * Copyright (C) 2007-2014 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/// @file
/// Declarations for MetricsHandler.
/// This file contains declarations for MetricsHandler, a dispatch handler class
/// used to collect and publish %FsBroker metrics.

#include <Common/Compat.h>

#include "MetricsHandler.h"

#include <AsyncComm/Comm.h>

#include <Common/Error.h>
#include <Common/Logger.h>
#include <Common/Properties.h>

#include <mutex>

using namespace Hypertable;
using namespace FsBroker;
using namespace std;

namespace {
  mutex g_mutex {};
  Comm *g_comm {};
}


MetricsHandler::MetricsHandler(PropertiesPtr &props, const std::string &type)
  : m_type(type) {
  int16_t port = props->get_i16("Hypertable.Metrics.Ganglia.Port");
  m_ganglia_collector = std::make_shared<MetricsCollectorGanglia>("fsbroker", port);
  m_collection_interval = props->get_i32("Hypertable.Monitoring.Interval");
  m_last_timestamp = Hypertable::get_ts64();
  {
    lock_guard<mutex> lock(g_mutex);
    int error;
    g_comm = Comm::instance();
    if ((error = g_comm->set_timer(m_collection_interval, this)) != Error::OK)
      HT_FATALF("Problem setting timer - %s", Error::get_text(error));
  }
}

MetricsHandler::~MetricsHandler() {
  lock_guard<mutex> lock(g_mutex);
  g_comm->cancel_timer(this);
  g_comm = 0;
}


void MetricsHandler::handle(Hypertable::EventPtr &event) {
  lock_guard<mutex> lock(g_mutex);
  int error;

  if (g_comm == 0)
    return;

  if (event->type == Hypertable::Event::TIMER) {

    int64_t now = Hypertable::get_ts64();
    m_metrics_process.collect(now, m_ganglia_collector.get());
    int64_t elapsed_millis = (now - m_last_timestamp) / 1000000LL;
    int64_t elapsed_seconds = elapsed_millis / 1000;
    m_ganglia_collector->update("type", m_type);

    {
      lock_guard<mutex> lock(m_mutex);

      m_ganglia_collector->update("errors", m_errors);

      int32_t avgSyncLatency = (m_syncs > 0) ? m_sync_latency/m_syncs : 0;
      m_ganglia_collector->update("syncLatency", avgSyncLatency);

      if (elapsed_millis > 0) {
        double sps = (double)m_syncs / (double)elapsed_seconds;
        m_ganglia_collector->update("syncs", sps);
        int64_t mbps = (m_bytes_read / 1000000) / elapsed_seconds;
        m_ganglia_collector->update("readThroughput", (int)mbps);
        mbps = (m_bytes_written / 1000000) / elapsed_seconds;
        m_ganglia_collector->update("writeThroughput", (int)mbps);
      }

      m_last_timestamp = now;
      m_errors = 0;
      m_syncs = 0;
      m_sync_latency = 0;
      m_bytes_read = 0;
      m_bytes_written = 0;
    }

    try {
      m_ganglia_collector->publish();
    }
    catch (Exception &e) {
      HT_INFOF("Problem publishing Ganglia metrics - %s", e.what());
    }

    if ((error = g_comm->set_timer(m_collection_interval, this)) != Error::OK)
      HT_FATALF("Problem setting timer - %s", Error::get_text(error));

  }
  else
    HT_FATALF("Unrecognized event - %s", event->to_str().c_str());

}
